src/bbd/geocoder/geocoder.py

```diff
@@ -0,0 +1,582 @@
+        If None will run all batches.
+
+    reset()
+        Method to reset progress and overwrite file at path.
+        WARNING: This method will overwrite file at path, do not use
+        unless you want to restart geocoding or if path is a path to 
+        existing file.
+
+    """
+
+    _cache_size = 4000
+    def __init__(self, data, email, path, batch_size = 3600,
+                 defaults = {"Country":"United States"}, 
+                 keep_index = True, index_name = ""):
+        """Constructor for bbd.geocoder.GeocodeLocations"""
+
+        #Creates geocoder with email passed as user_agent
+        self.geocoder = email #self.email is alias for self.geocoder
+        self.__real_geocoder = self._geocoder #for testing
+
+        self.defaults = defaults.copy()
+
+        self.data = pd.DataFrame(data)
+
+        if self.data.empty:
+            raise ValueError("Data cannot be empty!")
+
+        if not keep_index:
+            self.data = self.data.reset_index(drop=True)
+
+        try:
+            self.index_name = index_name or self.data.index.name
+        except AttributeError:
+            self.index_name = ""
+
+        #Verifying address components.
+        if len(self.data.columns) > 1:
+            #Components of address that will be passed to Nominatim.
+            component_list = ['address', 'street', 'city', 'county', 
+                              'state', 'country', 'zip5', 'zip', 
+                              'postal', 'postalcode']
+
+            self.included_cols = [col for col in self.data.columns
+                                  if col.lower() in component_list]
+
+            if not self.included_cols:
+                raise ValueError("If data a pd.DataFrame with multiple "\
+                                 "columns or a list of dictionaries, it "\
+                                 "must include labels/keys for at least "\
+                                 "some components of an address.")
+
+            #Defaults must be dict of address components.
+            if not all([key.lower() in component_list for key in defaults]):
+                raise ValueError("All keys in defaults must be components"\
+                                 "of an address.")
+        else:
+            self.included_cols = [self.data.columns[0]]
+
+        self.batch_size = batch_size
+
+        #Check file status then load or create file.
+        self.path = Path(path).resolve()
+        try:
+            #Load already geocoded results.
+            self.locations = pd.read_csv(self.path, sep = '\t',
+                                         index_col = self.index_name or 0)
+            
+            #check that self.locations is appropriate file.
+            if not all([col in self.locations.columns for col 
+                       in ["latitude", 'longitude', 'address']]):
+                raise ValueError(f"Path {self.path} is an unrelated file!")
```

Comment:
Maybe... but I want to be clear that the problem is the file is not being created/formatted by GeocodeLocations. I don't want to imply users can start with a file from elsewhere that has those columns and get GeocodeLocations to add on to it since I don't think the result would be stable.

Reply:
Maybe something like `"Path {self.path} is unrelated file. File should be generated by this program and contain columns for latitude, longitude, and address."`